File: Practice/AllDomains/CoreCS/Algorithms/Strings/BearAndSteadyGene.cs
using System;
using System.Collections.Generic;
using System.IO;
class Solution 
{
    /*	author:@shivkrthakur  */
    static void Main(String[] args) 
	{
        /* Enter your code here. Read input from STDIN. Print output to STDOUT. Your class should be named Solution */
		int n = Convert.ToInt32(Console.ReadLine());
		string s = Console.ReadLine().Trim();
        Dictionary<char,int> di = new Dictionary<char,int>() { 	{ 'A', 0 },	{ 'C', 0 },	{ 'G', 0 },	{ 'T', 0 } };
		int charCount = n / 4;
		int left = 0, right = n - 1, min = Int32.MaxValue;
		while(left < n)
		{
			di[s[left]]++; 
			if(di[s[left]] > charCount)
			{
				di[s[left]]--;
				left--;
				break;
			} 
			left++;
		}
		while(left >= 0 && right >= 0 && left < right)
		{
			di[s[right]]++;
			while((di['A'] > charCount || di['C'] > charCount || di['G'] > charCount || di['T'] > charCount) && left >= 0)
				di[s[left--]]--;
			min = Math.Min(min, Math.Max(0,right - left - 1));
			right--;
		}
		Console.WriteLine(min == Int32.MaxValue ? 0 : min);
    }
}

File: Practice/AllDomains/CoreCS/Algorithms/Strings/BearAndSteadyGene.js
/*	author:@shivkrthakur  */
var bearAndSteadyGene = function(input){
	var len = input.length;
	var limit = len/4;
	var overSize = 0;
	var overChars = { 'A': 0, 'C': 0, 'G': 0, 'T': 0 };
	var curChars = { 'A': 0, 'C': 0, 'G': 0, 'T': 0 };
	for(var i = 0; i < len; i++){
	  overChars[input[i]]++;
	  if(overChars[input[i]] > limit) overSize++;
	}

	Object.keys(overChars).forEach(function(key){
	   overChars[key] = Math.max(overChars[key] - limit, 0);  
	});

	var left = 0;
	while(left < overSize) curChars[input[left++]]++;

	var right = left;
	var output = Infinity;
	for(left = 0; left <= len - overSize && right < len;){
		if(curChars['A'] >= overChars['A'] && curChars['C'] >= overChars['C'] && curChars['G'] >= overChars['G'] && curChars['T'] >= overChars['T']){
			output = Math.min(output, (curChars['A'] + curChars['C'] + curChars['G'] + curChars['T']));
			curChars[input[left++]]--;
		} 
		else curChars[input[right++]]++;
	}          
	console.log(output);
}

function processData(input) {
    var arr = input.split('\n');
    bearAndSteadyGene(arr[1]);
} 

process.stdin.resume();
process.stdin.setEncoding("ascii");
_input = "";
process.stdin.on("data", function (input) {
    _input += input;
});

process.stdin.on("end", function () {
   processData(_input);
});
